### PR TITLE
extensions refactor to use span and std::shared_ptr

### PR DESCRIPTION
**Pull request by aldenml**

### CODE REVIEW

File: include/libtorrent/extensions.hpp [hunk 1]
```diff
@@ -158,12 +158,11 @@ POSSIBILITY OF SUCH DAMAGE.
...
-#include "libtorrent/sha1_hash.hpp" // for sha1_hash
-#include "libtorrent/session_handle.hpp"
-#include "libtorrent/peer_connection_handle.hpp"
-#include "libtorrent/span.hpp"
-#include "libtorrent/string_view.hpp"
+
+#include <libtorrent/config.hpp>
+#include <libtorrent/span.hpp>
+#include <libtorrent/string_view.hpp>
+#include <libtorrent/socket.hpp>
```
Comment (arvidn) [inline include/libtorrent/extensions.hpp]: I would expect "" would be more appropriate here, rather than <>

File: test/test_session_params.cpp [hunk 2]
```diff
@@ -58,7 +58,7 @@ namespace
 
 	struct custom_plugin : plugin
 	{
-		void added(session_handle h) override
+		void added(session_handle& h) override
```
Comment (arvidn) [inline test/test_session_params.cpp]: I thought session_handle was essentially a shared_ptr<> to a session_impl. why is it passed by reference? if it shouldn't be passed by value, I imagine it should be at least const reference.

File: include/libtorrent/extensions.hpp [hunk 3]
```diff
@@ -226,7 +228,7 @@ namespace libtorrent
 		{ return boost::shared_ptr<torrent_plugin>(); }
 
 		// called when plugin is added to a session
-		virtual void added(session_handle) {}
+		virtual void added(session_handle&) {}
```
Comment (arvidn) [inline include/libtorrent/extensions.hpp]: I imagine you meant to say const reference here, no

File: include/libtorrent/extensions.hpp (hunk 1 above)
Comment (aldenml) [inline include/libtorrent/extensions.hpp]: for some reason (probably copy&paste) it's `<>` in `dht_storage.hpp`, I will change it in both

File: test/test_session_params.cpp (hunk 2 above)
Comment (aldenml) [inline test/test_session_params.cpp]: I wanted to change it to keep it similar (in style) to when `peer_connection_handle` is passed. I will put the `const`.

File: include/libtorrent/extensions.hpp (hunk 1 above)
Comment (arvidn) [inline include/libtorrent/extensions.hpp]: yeah, I've sort of had this old idea of separating the DHT implementation from libtorrent, so I believe libtorrent includes are <> inside the DHT code

**Comment by arvidn** [comment]
lgtm

File: include/libtorrent/extensions.hpp (hunk 1 above)
Comment (aldenml) [inline include/libtorrent/extensions.hpp]: ok...I'm about to push the change to `std::shared_ptr`, finally found the issue (not changing boost::enable_shared_from_this<>)

File: include/libtorrent/bt_peer_connection.hpp [hunk 4]
```diff
@@ -192,10 +194,10 @@ namespace libtorrent
 
 #ifndef TORRENT_DISABLE_EXTENSIONS
 		bool supports_holepunch() const { return m_holepunch_id != 0; }
-		void set_ut_pex(boost::shared_ptr<ut_pex_peer_store> ut_pex)
-		{ m_ut_pex = ut_pex; }
+		void set_ut_pex(std::weak_ptr<ut_pex_peer_store> ut_pex)
+		{ m_ut_pex.swap(ut_pex); }
```
Comment (arvidn) [inline include/libtorrent/bt_peer_connection.hpp]: I would expect a: `m_ut_pex = std::move(ut_pex);` to be more appropriate

File: include/libtorrent/bt_peer_connection.hpp [hunk 5]
```diff
@@ -457,9 +459,9 @@ namespace libtorrent
 		// 0 if not supported
 		std::uint8_t m_share_mode_id;
 
-		boost::shared_ptr<ut_pex_peer_store> m_ut_pex;
+		std::weak_ptr<ut_pex_peer_store> m_ut_pex;
```
Comment (arvidn) [inline include/libtorrent/bt_peer_connection.hpp]: good catch

**Pull request merged**